Scroll-to-text-fragment is leaked to the loaded page through Navigation Timing.

megangardner opened the pull request:
#### 1d776f6172680f23acb01d5abca5a41722f2c7df
<pre>
Scroll-to-text-fragment is leaked to the loaded page through Navigation Timing.
<a href="https://bugs.webkit.org/show_bug.cgi?id=276925">https://bugs.webkit.org/show_bug.cgi?id=276925</a>
<a href="https://rdar.apple.com/124717009">rdar://124717009</a>

Reviewed by Ryosuke Niwa.

We need to strip the fragment from the URL earlier so that the full URL does not end up
in the performance metrics, leaking the fragment that should be invisible to the page.

WPT test will need to be upstreamed after this patch lands.

* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-target.html:
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment.html:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setURL):
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::startLoadingMainResource):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setMainFrameURLFragment):
* Source/WebCore/page/Page.h:
(WebCore::Page::mainFrameURLFragment const):

Canonical link: <a href="https://commits.webkit.org/281273@main">https://commits.webkit.org/281273@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b0897c3e4258f15232b02c60ffa33f54dee9c605

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59333 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38676 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11853 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63247 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9776 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61462 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46330 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10007 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48184 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6949 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61363 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36131 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51352 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29008 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32842 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8607 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8780 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54800 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8887 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64980 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3261 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8812 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55528 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3272 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51349 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55620 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13148 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2711 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34492 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35575 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36661 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35320 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/page/Page.cpp:
```diff
@@ -756,6 +756,12 @@ void Page::setMainFrameURL(const URL& url)
 
+void Page::setMainFrameURLFragment(String&& fragment)
+{
+    if (!fragment.isEmpty() && !fragment.isNull())
```
`isEmpty()` also checks `isNull()`,  so`&& !fragment.isNull()` can be removed

rr-codes commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -2140,7 +2140,14 @@ void DocumentLoader::startLoadingMainResource()
+    m_request.setURL(url, m_request.didFilterLinkDecoration());
+    auto page = m_frame->protectedPage();
+    if (m_frame && page && !fragmentDirective.isEmpty() && !fragmentDirective.isNull())
+        page->setMainFrameURLFragment(WTFMove(fragmentDirective)); // <MMG> protected page?
```
remove comment

rr-codes commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -2140,7 +2140,14 @@ void DocumentLoader::startLoadingMainResource()
+    auto url = m_request.url();
+    auto fragmentDirective = url.consumeFragmentDirective();
+    m_request.setURL(url, m_request.didFilterLinkDecoration());
+    auto page = m_frame->protectedPage();
```
```suggestion
    Ref (or RefPtr) page = m_frame->protectedPage();
```

rr-codes commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -2140,7 +2140,14 @@ void DocumentLoader::startLoadingMainResource()
+    auto fragmentDirective = url.consumeFragmentDirective();
+    m_request.setURL(url, m_request.didFilterLinkDecoration());
+    auto page = m_frame->protectedPage();
+    if (m_frame && page && !fragmentDirective.isEmpty() && !fragmentDirective.isNull())
```
`m_frame` is guaranteed to be non-null here, assuming it is non-null in `auto page = m_frame->protectedPage();`

rr-codes replied:
ditto about the isEmpty/isNull thing

rniwa commented on LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-target.html:
```diff
@@ -42,6 +42,7 @@
   let results = {
     scrollPosition: position,
     href: window.location.href,
+    performanceEntry: performance.getEntries()[0].name,
```
Is this change done in the upstream?
We should mention that in the commit message.

megangardner replied:
No, we need to upstream it when this is pushed, which I understand is the easier direction?

rniwa commented on Source/WebCore/page/Page.h:
```diff
@@ -360,6 +360,8 @@ class Page : public RefCounted<Page>, public Supplementable<Page>, public CanMak
     SecurityOrigin& mainFrameOrigin() const;
     WEBCORE_EXPORT void setMainFrameURL(const URL&);
+    WEBCORE_EXPORT void setMainFrameURLFragment(String&&);
+    String mainFrameURLFragment() const { return m_mainFrameURLFragment; }
```
So text fragment will never be used in an iframe?

megangardner replied:
https://wicg.github.io/scroll-to-text-fragment/#motivation
Yes, they are top level navigation only.

rniwa commented on Source/WebCore/page/Page.cpp:
```diff
@@ -756,6 +756,12 @@ void Page::setMainFrameURL(const URL& url)
 
+void Page::setMainFrameURLFragment(String&& fragment)
+{
+    if (!fragment.isEmpty())
```
The caller already checks it. There is no need to check it again.
Add an assert instead maybe?

aprotyas replied:
> The caller already checks it. There is no need to check it again.

It might be more ergonomic to move the emptiness check in here and not have the caller do it, assuming more callers in the future.

megangardner replied:
Lots of this is left over from when I was trying to figure out why it was crashing, I'll cull it all down.

megangardner committed:
Scroll-to-text-fragment is leaked to the loaded page through Navigation Timing.
https://bugs.webkit.org/show_bug.cgi?id=276925
rdar://124717009

Reviewed by Ryosuke Niwa.

We need to strip the fragment from the URL earlier so that the full URL does not end up
in the performance metrics, leaking the fragment that should be invisible to the page.

WPT test will need to be upstreamed after this patch lands.

* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-target.html:
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment.html:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setURL):
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::startLoadingMainResource):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setMainFrameURLFragment):
* Source/WebCore/page/Page.h:
(WebCore::Page::mainFrameURLFragment const):

Canonical link: https://commits.webkit.org/281273@main
